File: AoC2020-Python/day05-binary-boarding.py
### ADVENT OF CODE 2020 DAY 5 BINARY BOARDING ###


import helper.helper as helper


## * puzzle part 1
def puzzle_part_1(puzzle_input):
    # replace characters with binary equivalent
    binary_puzzle_input = [
        x.replace("F", "0").replace("B", "1").replace("L", "0").replace("R", "1")
        for x in puzzle_input
    ]

    # convert from string to base 2 (binary) int
    seats = [int(x, 2) for x in binary_puzzle_input]

    # sort to put largest seat id at end of list
    seats.sort()

    # return answer
    return seats[-1], seats


## * puzzle part 2
def puzzle_part_2(seat_list):
    # find seat missing from list in puzzle 1
    my_seat = 0
    seats = seat_list
    for i in range(len(seats)):
        if seats[i + 1] - seats[i] != 1:
            my_seat = seats[i] + 1
            break

    # return answer
    return my_seat


## * main function
def main():
    # print title
    helper.print_title("2020", "5", "Binary Boarding")

    # get puzzle input
    puzzle_input = helper.load_puzzle_input("inputs/day05/pi.txt")

    # solve part 1 and print QA
    answer, seat_list = puzzle_part_1(puzzle_input)
    helper.print_question_answer(
        "What is the highest seat ID on a boarding pass? ", answer,
    )

    # solve part 2 and print QA
    answer = puzzle_part_2(seat_list)
    helper.print_question_answer(
        "What is the ID of your seat? ", answer,
    )


if __name__ == "__main__":
    main()

File: AoC2020-C++/day05-binary-boarding.cpp
// * ADVENT OF CODE 2020 - DAY 5 - BINARY BOARDING *

#include <algorithm>

#include "helper/helper.h"

// * puzzle part 2 function
int puzzlePart2(std::vector<int> puzzleVector) {
    // seat tracking var
    int mySeat = 0;

    // loop through each element in vector
    for (unsigned long int i = 0; i < puzzleVector.size(); i++) {
        // check to see if the next element minus current element does not equal 1
        // if so, we have a missing value which should be our seat
        if ((puzzleVector[i + 1] - puzzleVector[i]) != 1) {
            mySeat = puzzleVector[i] + 1;
            break;
        }
    }

    // return answer
    return mySeat;
}

// * binary replace characters, convert to base 10, sort
std::vector<int> vecBinaryReplaceConvertSort(std::vector<std::string> puzzleVector) {
    // replace characters with binary equivalent
    for (unsigned long int i = 0; i < puzzleVector.size(); i++) {
        std::replace(puzzleVector[i].begin(), puzzleVector[i].end(), 'F', '0');
        std::replace(puzzleVector[i].begin(), puzzleVector[i].end(), 'B', '1');
        std::replace(puzzleVector[i].begin(), puzzleVector[i].end(), 'L', '0');
        std::replace(puzzleVector[i].begin(), puzzleVector[i].end(), 'R', '1');
    }

    // convert from string to base 2 (binary) int
    std::vector<int> puzzleVectorInt;
    for (auto line : puzzleVector) {
        puzzleVectorInt.push_back(std::stoi(line, nullptr, 2));
    }

    // sort to put largest seat id at end of list
    sort(puzzleVectorInt.begin(), puzzleVectorInt.end());

    // return new converted vector
    return puzzleVectorInt;
}

// * main function
int main() {
    // print title
    helper::printTitle("2020", "5", "Binary Boarding");

    // load puzzle input
    std::vector<std::string> puzzleVector = helper::loadPuzzleInputStringVec("inputs/day05/pi.txt");

    // setup converted base 10 vector
    std::vector<int> puzzleVectorInt = vecBinaryReplaceConvertSort(puzzleVector);

    // solve part 1
    int answer = puzzleVectorInt.back();

    // print question and answer
    helper::printQuestionAnswer("\nWhat is the highest seat ID on a boarding pass? ", std::to_string(answer));

    // solve part 2
    answer = puzzlePart2(puzzleVectorInt);

    // print question and answer
    helper::printQuestionAnswer("What is the ID of your seat? ", std::to_string(answer));

    return 0;
}